src/backend/winit.rs

```diff
@@ -0,0 +1,652 @@
+//! Implementation of backend traits for types provided by `winit`
+
+use backend::{SeatInternal, TouchSlotInternal};
+use backend::graphics::GraphicsBackend;
+use backend::graphics::egl::{CreationError, EGLContext, EGLGraphicsBackend, GlAttributes, Native,
+                             PixelFormat, PixelFormatRequirements, Robustness, SwapBuffersError};
+use backend::input::{Axis, AxisSource, Event as BackendEvent, InputBackend, InputHandler, KeyState,
+                     KeyboardKeyEvent, MouseButton, MouseButtonState, PointerAxisEvent, PointerButtonEvent,
+                     PointerMotionAbsoluteEvent, Seat, SeatCapabilities, TouchCancelEvent, TouchDownEvent,
+                     TouchMotionEvent, TouchSlot, TouchUpEvent, UnusedEvent};
+use nix::c_void;
+
+use std::cmp;
+use std::error::Error;
+use std::fmt;
+use std::rc::Rc;
+use wayland_client::egl as wegl;
+use winit::{CreationError as WinitCreationError, ElementState, Event, EventsLoop,
+            MouseButton as WinitMouseButton, MouseCursor, MouseScrollDelta, Touch, TouchPhase, Window,
+            WindowBuilder, WindowEvent};
+use winit::os::unix::{WindowExt, get_x11_xconnection};
+
+/// Window with an active EGL Context created by `winit`. Implements the
+/// `EGLGraphicsBackend` graphics backend trait
+pub struct WinitGraphicsBackend {
+    window: Rc<Window>,
+    context: EGLContext,
+}
+
+/// Abstracted event loop of a `winit` `Window` implementing the `InputBackend` trait
+///
+/// You need to call `dispatch_new_events` periodically to receive any events.
+pub struct WinitInputBackend {
+    events_loop: EventsLoop,
+    window: Rc<Window>,
+    surface: Option<wegl::WlEglSurface>,
+    time_counter: u32,
+    key_counter: u32,
+    seat: Seat,
+    input_config: (),
+    handler: Option<Box<InputHandler<WinitInputBackend> + 'static>>,
+}
+
+/// Create a new `WinitGraphicsBackend`, which implements the `EGLGraphicsBackend`
+/// graphics backend trait and a corresponding `WinitInputBackend`, which implements
+/// the `InputBackend` trait
+pub fn init() -> Result<(WinitGraphicsBackend, WinitInputBackend), CreationError> {
+    init_from_builder(WindowBuilder::new())
+}
+
+/// Create a new `WinitGraphicsBackend`, which implements the `EGLGraphicsBackend`
+/// graphics backend trait, from a given `WindowBuilder` struct and a corresponding
+/// `WinitInputBackend`, which implements the `InputBackend` trait
+pub fn init_from_builder(builder: WindowBuilder)
+                         -> Result<(WinitGraphicsBackend, WinitInputBackend), CreationError> {
+    init_from_builder_with_gl_attr(builder,
+                                   GlAttributes {
+                                       version: None,
+                                       profile: None,
+                                       debug: cfg!(debug_assertions),
+                                       robustness: Robustness::TryRobustLoseContextOnReset,
```

Comment:
This causes context creation to fail, because apparently my EGL implementation reports, that it supports robustness, but creation later fails with `BadAttribute`, if it is set to any other value then `NotRobust`.